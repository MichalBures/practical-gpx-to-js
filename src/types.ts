export interface GpxWaypoint {
	lat: number;
	lon: number;
	time: Date | null;
	name: string | null;
	description: string | null;
	symbol: string | null;
	altitude: number | null;
}

export interface GpxTrackpoint {
	lat: number;
	lon: number;
	time: Date | null;
	altitude: number | null;
	speed: number | null;
	cadence: number | null;
	heartRate: number | null;
}

export interface GpxTrack {
	name: string | null;
	trackpoints: Array<GpxTrackpoint>;
	segments: Array<number>;
}

export interface GpxMetadata {
	name: string | null;
	description: string | null;
	creator: string | null;
	time: Date | null;
}

export interface Gpx {
	metadata: GpxMetadata;
	waypoints: Array<GpxWaypoint> | null;
	tracks: Array<GpxTrack> | null;
}

export interface ParsedGpx {
	gpx: {
		$: {
			creator: string;
			[propName: string]: any;
		};
		metadata: Array<any>;
		trk: Array<any>;
		wpt: Array<any>;
	};
}
