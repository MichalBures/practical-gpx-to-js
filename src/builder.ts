import * as GpxTypes from './types';
import { Builder } from 'xml2js';

export function buildGpx({
	metadata,
	tracks = [],
	waypoints = [],
}: {
	metadata: Partial<GpxTypes.GpxMetadata>;
	tracks?: Array<Partial<GpxTypes.GpxTrack>>;
	waypoints?: Array<Partial<GpxTypes.GpxWaypoint>>;
}) {
	let trks = tracksToGpxTrk(tracks);
	let wpts = waypointsToGpxWpt(waypoints);

	var builder = new Builder();

	let gpx: any = {
		gpx: {
			$: {
				version: '1.1',
				xmlns: 'http://www.topografix.com/GPX/1/1',
				creator: metadata ? metadata.creator : '',
			},
			metadata: {
				time: metadata.time
					? metadata.time.toISOString()
					: new Date().toISOString(),
			},
		},
	};

	if (metadata.name) {
		gpx.gpx.metadata.name = metadata.name;
	}
	if (metadata.description) {
		gpx.gpx.metadata.desc = metadata.description;
	}

	if (wpts.length > 0) {
		gpx.gpx.wpt = wpts;
	}

	if (trks.length > 0) {
		gpx.gpx.trk = trks;
	}

	var xml = builder.buildObject(gpx);

	return xml;
}

export function waypointsToGpxWpt(
	waypoints: Array<Partial<GpxTypes.GpxWaypoint>> = []
) {
	let wpts: Array<any> = [];

	waypoints.forEach((waypoint) => {
		let wpt = waypointToGpxWpt(waypoint);
		wpts.push(wpt);
	});

	return wpts;
}

export function waypointToGpxWpt(waypoint: Partial<GpxTypes.GpxWaypoint>) {
	let wpt: any = {
		$: {
			lat: waypoint.lat ? waypoint.lat.toString(10) : undefined,
			lon: waypoint.lon ? waypoint.lon.toString(10) : undefined,
		},
	};

	if (waypoint.name) {
		wpt.name = waypoint.name;
	}

	if (waypoint.description) {
		wpt.desc = waypoint.description;
	}

	if (waypoint.time) {
		wpt.time = waypoint.time.toISOString();
	}

	if (waypoint.symbol) {
		wpt.sym = waypoint.symbol;
	}

	if (waypoint.altitude) {
		wpt.ele = waypoint.altitude;
	}

	return wpt;
}

export function tracksToGpxTrk(tracks: Array<Partial<GpxTypes.GpxTrack>> = []) {
	let trks: Array<any> = [];

	tracks.forEach((track) => {
		let trk = trackToGpxTrk(track);
		trks.push(trk);
	});

	return trks;
}

export function trackpointToGpxTrkpt(
	trackpoint: Partial<GpxTypes.GpxTrackpoint>
) {
	let point: any = {
		trkpt: {
			$: {
				lat: trackpoint.lat ? trackpoint.lat.toString(10) : undefined,
				lon: trackpoint.lon ? trackpoint.lon.toString(10) : undefined,
			},
		},
	};

	if (trackpoint.time) {
		point.trkpt.time = trackpoint.time.toISOString();
	}

	if (trackpoint.altitude) {
		point.trkpt.ele = trackpoint.altitude;
	}

	if (trackpoint.speed) {
		point.trkpt.speed = trackpoint.speed;
	}

	if (trackpoint.heartRate == null && trackpoint.cadence == null) {
		return point;
	}

	// Add extensions

	let extension: any = {};

	if (trackpoint.heartRate != null) {
		extension['gpxtpx:hr'] = [trackpoint.heartRate.toString()];
	}
	if (trackpoint.cadence != null) {
		extension['gpxtpx:cad'] = [trackpoint.cadence.toString()];
	}

	let extensions = [
		{
			'gpxtpx:TrackPointExtension': [extension],
		},
	];

	point.trkpt.extensions = extensions;

	return point;
}

export function trackToGpxTrk(track: Partial<GpxTypes.GpxTrack>) {
	let trk: any = {
		name: track.name,
		trkseg: [],
	};

	// TS is complaining about undefined in the forEach below
	const segments = track.segments;

	if (
		track.trackpoints == undefined ||
		track.trackpoints.length === 0 ||
		segments == undefined ||
		segments.length === 0
	) {
		return trk;
	}

	let trackSegmentPoints: Array<any> = [];
	let trackSegmentIndex = 0;
	let trackPointCount = 0;

	track.trackpoints.forEach((trackpoint) => {
		let trkpt = trackpointToGpxTrkpt(trackpoint);
		trackSegmentPoints.push(trkpt);

		trackPointCount++;

		if (trackPointCount >= segments[trackSegmentIndex]) {
			trackPointCount = 0;
			trackSegmentIndex++;

			// Add track segment to track
			trk.trkseg.push(trackSegmentPoints);
			trackSegmentPoints = [];

			// Error: There are more segments than actual track points
			if (trackSegmentIndex >= segments.length) {
				return false;
			}
		}

		return undefined;
	});

	return trk;
}
