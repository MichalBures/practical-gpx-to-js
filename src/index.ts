export { parseGpx } from './parser';
export { buildGpx } from './builder';
export * from './types';
