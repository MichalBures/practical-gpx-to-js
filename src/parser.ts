import * as GpxTypes from './types';
import { parseStringPromise } from 'xml2js';
import { parseISO } from 'date-fns';

export async function parseGpx(gpxString: string): Promise<GpxTypes.Gpx> {
	let parsedGpx = await parseGpxString(gpxString);

	let metadata = parseGpxMetadata(parsedGpx);
	let waypoints = parseGpxWpt(parsedGpx);
	let tracks = parseGpxTrk(parsedGpx);

	return {
		metadata,
		waypoints,
		tracks,
	};
}

/**
 * Parses .gpx file using xml2js into js structures
 *
 * @param gpxString .gpx file contents
 */
export async function parseGpxString(
	gpxString: string
): Promise<GpxTypes.ParsedGpx> {
	let parsedGpx;

	try {
		parsedGpx = await parseStringPromise(gpxString);
	} catch (e) {
		throw `Couldn't parse given .gpx, error: ${e}`;
	}

	if (parsedGpx.gpx === undefined) {
		throw 'Given file is not a .gpx file';
	}

	return parsedGpx;
}

export function hasGpxTagValue(value: any): value is Array<any> {
	return value != undefined && Array.isArray(value) && value.length > 0;
}

export function parseGpxMetadata(parsedGpx: GpxTypes.ParsedGpx) {
	let metadata: GpxTypes.GpxMetadata = {
		creator: null,
		name: null,
		description: null,
		time: null,
	};
	// <gpx>
	if (parsedGpx.gpx.$ !== undefined && parsedGpx.gpx.$.creator !== undefined) {
		metadata.creator = parsedGpx.gpx.$.creator;
	}

	// <metadata>
	if (hasGpxTagValue(parsedGpx.gpx.metadata)) {
		let gpxMetadata = parsedGpx.gpx.metadata[0];
		metadata.name = hasGpxTagValue(gpxMetadata.name)
			? gpxMetadata.name[0]
			: null;
		metadata.description = hasGpxTagValue(gpxMetadata.desc)
			? gpxMetadata.desc[0]
			: null;
		metadata.time = hasGpxTagValue(gpxMetadata.time)
			? parseISO(gpxMetadata.time[0])
			: null;
	}

	return metadata;
}

export function parseGpxWpt(
	parsedGpx: GpxTypes.ParsedGpx
): Array<GpxTypes.GpxWaypoint> | null {
	let waypoints: Array<GpxTypes.GpxWaypoint> = [];

	if (!hasGpxTagValue(parsedGpx.gpx.wpt)) {
		return null;
	}

	parsedGpx.gpx.wpt.forEach((wpt) => {
		// Skip waypoints without both Lat and Lon
		if (
			wpt.$ == undefined ||
			wpt.$.lat == undefined ||
			wpt.$.lon == undefined
		) {
			return;
		}

		let waypoint: Partial<GpxTypes.GpxWaypoint> = {};

		waypoint.lat = Number(wpt.$.lat);
		waypoint.lon = Number(wpt.$.lon);
		waypoint.time = hasGpxTagValue(wpt.time) ? parseISO(wpt.time[0]) : null;
		waypoint.name = hasGpxTagValue(wpt.name) ? wpt.name[0] : null;
		waypoint.description = hasGpxTagValue(wpt.desc) ? wpt.desc[0] : null;
		waypoint.symbol = hasGpxTagValue(wpt.sym) ? wpt.sym[0] : null;
		waypoint.altitude = hasGpxTagValue(wpt.ele) ? Number(wpt.ele[0]) : null;

		waypoints.push(<GpxTypes.GpxWaypoint>waypoint);
	});

	return waypoints;
}

export function parseGpxTrk(
	parsedGpx: GpxTypes.ParsedGpx
): Array<GpxTypes.GpxTrack> | null {
	let tracks: Array<GpxTypes.GpxTrack> = [];

	if (!hasGpxTagValue(parsedGpx.gpx.trk)) {
		return null;
	}

	parsedGpx.gpx.trk.forEach((trk) => {
		let track: GpxTypes.GpxTrack = {
			name: hasGpxTagValue(trk.name) ? trk.name[0] : null,
			trackpoints: [],
			segments: [],
		};

		// Track has no trackpoints
		if (!trk.trkseg || !Array.isArray(trk.trkseg)) {
			tracks.push(track);
			return false;
		}

		let trkSegments: Array<{ trkpt: Array<any> }> = trk.trkseg;

		trkSegments.forEach((segment) => {
			track.segments.push(segment.trkpt.length);

			let trkPoints: Array<any> = segment.trkpt;

			trkPoints.forEach((point) => {
				// Skip trackpoints without both Lat and Lon
				if (
					point.$ == undefined ||
					point.$.lat == undefined ||
					point.$.lon == undefined
				) {
					return;
				}

				let trackpoint: Partial<GpxTypes.GpxTrackpoint> = {
					cadence: null,
					heartRate: null,
				};

				trackpoint.lat = Number(point.$.lat);
				trackpoint.lon = Number(point.$.lon);
				trackpoint.time = hasGpxTagValue(point.time)
					? parseISO(point.time[0])
					: null;
				trackpoint.altitude = hasGpxTagValue(point.ele)
					? Number(point.ele)
					: null;
				trackpoint.speed = hasGpxTagValue(point.speed)
					? Number(point.speed)
					: null;

				// Handle trackpoint extensions such as heart rate and cadence
				if (point.extensions != null && hasGpxTagValue(point.extensions)) {
					let extensions = parseTrackpointExtensions(point.extensions);

					if (extensions.cad !== undefined) {
						trackpoint.cadence = extensions.cad;
					}
					if (extensions.hr !== undefined) {
						trackpoint.heartRate = extensions.hr;
					}
				}

				track.trackpoints.push(<GpxTypes.GpxTrackpoint>trackpoint);
			});
		});

		tracks.push(track);
		return undefined;
	});

	return tracks;
}

/**
 * Parses .gpx trackpoint extension cadence and heart rate to Trackpoint values
 *
 * @param extensions
 */
export function parseTrackpointExtensions(
	extensions: Array<any>
): { hr?: number; cad?: number } {
	let extensionValues: any = {};

	// TODO: Seems there are many different extension, will handle when needed
	// extensions = [ { 'gpxtpx:TrackPointExtension': [ [Object] ] } ]
	extensions.forEach((extension) => {
		for (let extensionName in extension) {
			if (!extension.hasOwnProperty(extensionName)) {
				continue;
			}

			// Is it really a TrackPointExtension
			let extensionNameOnly = extensionName.split(':');
			if (
				extensionNameOnly.length !== 2 &&
				extensionNameOnly[1] !== 'TrackPointExtension'
			) {
				continue;
			}

			// It is, now get the values
			let trackPointExtensions = extension[extensionName];

			trackPointExtensions.forEach((trackPointExtension: any) => {
				for (let extensionValueName in trackPointExtension) {
					if (!trackPointExtension.hasOwnProperty(extensionValueName)) {
						continue;
					}

					// Now we have value name, such as cad, hr
					let valueNameOnly = extensionValueName.split(':');
					if (valueNameOnly.length !== 2) {
						continue;
					}

					// Pick only select extension values
					let valueName = '';
					if (valueNameOnly[1] === 'cad') {
						valueName = valueNameOnly[1];
						let extensionValue = trackPointExtension[extensionValueName];
						if (hasGpxTagValue(extensionValue)) {
							extensionValues[valueName] = Number(extensionValue[0]);
						}
					}
					if (valueNameOnly[1] === 'hr') {
						valueName = valueNameOnly[1];
						let extensionValue = trackPointExtension[extensionValueName];
						if (hasGpxTagValue(extensionValue)) {
							extensionValues[valueName] = Number(extensionValue[0]);
						}
					}
				}
			});
		}
	});

	return extensionValues;
}
