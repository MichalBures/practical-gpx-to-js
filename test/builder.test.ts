import * as fs from 'fs';
import * as path from 'path';
import {
	parseGpxString,
	parseGpxWpt,
	parseGpxTrk,
	parseGpxMetadata,
} from '../src/parser';
import { waypointsToGpxWpt, tracksToGpxTrk, buildGpx } from '../src/builder';
import { GpxTrack } from '../src/types';

describe('Builder', () => {
	it('build parsed .gpx', async () => {
		let file = fs.readFileSync(path.join(__dirname, 'Praha-88km.gpx'), {
			encoding: 'utf8',
		});

		let result: any = {};
		let parsedGpx: any = await parseGpxString(file);

		let metadata = parseGpxMetadata(parsedGpx);
		// console.log(metadata);

		let waypoints = parseGpxWpt(parsedGpx);
		// console.log(waypoints);

		if (waypoints !== null) {
			let wpts = waypointsToGpxWpt(waypoints);
			// console.log(wpts);
		}

		let tracks = parseGpxTrk(parsedGpx);
		// console.log(JSON.stringify(tracks, null, ' '));

		if (tracks !== null) {
			let trks = tracksToGpxTrk(tracks);
			// console.log(JSON.stringify(trks, null, ' '));
		}

		let firstLastTrackpointTrack: Partial<GpxTrack> = {};

		if (tracks != null && tracks.length > 0) {
			firstLastTrackpointTrack = { ...tracks[0] };
			firstLastTrackpointTrack.trackpoints = [
				tracks[0].trackpoints[0],
				tracks[0].trackpoints[tracks[0].trackpoints.length - 1],
			];
			firstLastTrackpointTrack.segments = [2];
		}

		console.log(firstLastTrackpointTrack);

		let gpx = {};
		if (waypoints != null && tracks !== null) {
			gpx = buildGpx({
				metadata,
				waypoints,
				tracks: [firstLastTrackpointTrack],
			});
			// console.log(gpx);
		}

		expect(gpx).toMatchSnapshot();
	});
});
