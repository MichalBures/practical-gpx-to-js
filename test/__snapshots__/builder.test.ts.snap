// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Builder build parsed .gpx 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" standalone=\\"yes\\"?>
<gpx version=\\"1.1\\" xmlns=\\"http://www.topografix.com/GPX/1/1\\" creator=\\"AllTrails https://www.alltrails.com/ - Praha-8,8km\\">
  <metadata>
    <time>2019-10-18T21:50:26.000Z</time>
    <name>Praha-8,8km</name>
  </metadata>
  <wpt lat=\\"50.088387\\" lon=\\"14.3919342\\">
    <name>Start</name>
    <time>2010-01-01T00:00:00.000Z</time>
    <ele>280</ele>
  </wpt>
  <wpt lat=\\"50.08826\\" lon=\\"14.39154\\">
    <name>End</name>
    <time>2010-01-01T00:32:55.000Z</time>
    <ele>282</ele>
  </wpt>
  <trk>
    <name>Praha-8,8km on AllTrails</name>
    <trkseg>
      <trkpt lat=\\"50.088387\\" lon=\\"14.3919342\\">
        <time>2010-01-01T00:00:00.000Z</time>
        <ele>280</ele>
      </trkpt>
      <trkpt lat=\\"50.08826\\" lon=\\"14.39154\\">
        <time>2010-01-01T00:32:55.000Z</time>
        <ele>282</ele>
      </trkpt>
    </trkseg>
  </trk>
</gpx>"
`;
